### PR TITLE
Faster TXGen

### PR DESCRIPTION
**Pull request by alajko**
## Issue

Current TXGen is not setup to send transactions to testnet. In order to send a volume of txns via txgen we need to sync state to current blockchain. This TXGen adds ability to send transactions directly to tesnet AND send large vollume of txns.

## Test

Running on internal testnet (block 24 onwards, 5:39 pm 12 April)

## TODO NEXT

- [ ] Test: Stability of TXGen over hours of sending a large vollume of transactions.
- [ ] Test: Stability of TXGen in sending a larger vollume of txns (>500) per batch.
- [ ] Archival Feature: Ability of TXGen to start and stop without having to sync entire blockchain everytime.
- [ ] Agility Feature: Agility of TXGen to syncto  current state quickly and generate valid txns while syncing to entire blockchain in background.
- [ ] Stability Feature: Stability of TXGen when it has to sync to a blockchain with height in 1000's.

### CODE REVIEW

**Comment by alajko** [comment]
Jenkins, test this please

**Comment by alajko** [comment]
Jenkins, test this please.

**Comment by alajko** [comment]
Freshly tested again on internat testnet (running from block 253 onwards (9:39 pm, 12 April)).

**Comment by alajko** [comment]
Jenkins, test this please.

File: cmd/client/txgen/main.go [hunk 1]
```diff
@@ -48,16 +60,15 @@ func main() {
...
-
+	//crossShardRatio := flag.Int("cross_shard_ratio", 30, "The percentage of cross shard transactions.")
+	shardIDFlag := flag.Int("shardID", 0, "The shardID the node belongs to.")
 	// Key file to store the private key
 	keyFile := flag.String("key", "./.txgenkey", "the private key file of the txgen")
 	flag.Var(&utils.BootNodes, "bootnodes", "a list of bootnode multiaddress")
-
 	flag.Parse()
-
+	time.Sleep(checkFrequency * time.Second) //This lets other nodes start their services before I query for blockchain height.
```
[the remark this answers is not in the document]
Comment (alajko) [reply cmd/client/txgen/main.go]: removed.

File: cmd/client/txgen/main.go [hunk 2]
```diff
@@ -86,141 +96,90 @@ func main() {
...
 	}
-	for _, shardID := range shardIDs {
-		node := node.New(host, &consensus.Consensus{ShardID: shardID}, nil, false)
-		// Assign many fake addresses so we have enough address to play with at first
-		nodes = append(nodes, node)
-	}
-
-	// Client/txgenerator server node setup
-	consensusObj, err := consensus.New(host, 0, p2p.Peer{}, nil)
+	consensus, err := consensus.New(myhost, uint32(shardID), p2p.Peer{}, nil)
```
[the remark this answers is not in the document]
Comment (alajko) [reply cmd/client/txgen/main.go]: done.

File: cmd/client/txgen/main.go [hunk 3]
```diff
@@ -86,141 +96,90 @@ func main() {
...
-		// wait for 3 seconds for client to send ping message to leader
-		// FIXME (leo) the readySignal should be set once we really sent ping message to leader
-		time.Sleep(3 * time.Second) // wait for nodes to be ready
-		for _, i := range shardIDs {
-			readySignal <- i
-		}
-	}()
-
-	// Transaction generation process
+	node := node.New(myhost, consensus, nil, true) //Changed it : no longer archival node.
```
[the remark this answers is not in the document]
Comment (alajko) [reply cmd/client/txgen/main.go]: Done.

File: node/node_syncing.go [hunk 4]
```diff
@@ -103,12 +103,16 @@ func (node *Node) SupportSyncing() {
...
 	go node.SendNewBlockToUnsync()
-
 	if node.NodeConfig.Role() != nodeconfig.ShardLeader && node.NodeConfig.Role() != nodeconfig.BeaconLeader {
 		go node.DoSyncing(node.blockchain, node.Worker, node.GetSyncingPeers, true)
 	}
 }
 
+// GetSync gets sync-ed to blockchain without joining consensus
+func (node *Node) GetSync() {
+	go node.DoSyncing(node.blockchain, node.Worker, node.GetBeaconSyncingPeers, false) //Don't join consensus
```
[the remark this answers is not in the document]
Comment (alajko) [reply node/node_syncing.go]: removed.

File: cmd/client/txgen/main.go [hunk 5]
```diff
@@ -48,16 +60,15 @@ func main() {
 
 	maxNumTxsPerBatch := flag.Int("max_num_txs_per_batch", 20000, "number of transactions to send per message")
 	logFolder := flag.String("log_folder", "latest", "the folder collecting the logs of this execution")
-	duration := flag.Int("duration", 10, "duration of the tx generation in second. If it's negative, the experiment runs forever.")
+	duration := flag.Int("duration", 0, "duration of the tx generation in second. If it's negative, the experiment runs forever.")
```
[the remark this answers is not in the document]
Comment (alajko) [reply cmd/client/txgen/main.go]: fixed.

File: cmd/client/txgen/main.go [hunk 6]
```diff
@@ -86,141 +96,90 @@ func main() {
 	if peerPubKey == nil {
 		panic(fmt.Errorf("generate key error"))
 	}
-
+	shardID := *shardIDFlag
+	var shardIDs []uint32
+	shardIDs = append(shardIDs, uint32(shardID))
```
[the remark this answers is not in the document]
Comment (alajko) [reply cmd/client/txgen/main.go]: Its needed for node.client

**Comment by LeoHChen** [comment]
discard this older PR.

**Pull request closed without merging**